File: app/src/main/java/moe/yuuta/mipushtester/push/MessageDetailBindingUtils.java
package moe.yuuta.mipushtester.push;

import android.text.method.ScrollingMovementMethod;
import android.widget.TextView;
import android.os.Build;

import com.elvishew.xlog.Logger;
import com.elvishew.xlog.XLog;
import com.google.gson.GsonBuilder;
import com.xiaomi.mipush.sdk.MiPushMessage;

import androidx.databinding.BindingAdapter;

import java.lang.reflect.Method;

import moe.yuuta.common.Constants;
import moe.yuuta.mipushtester.R;

public class MessageDetailBindingUtils {
    private static final Logger logger = XLog.tag(MessageDetailBindingUtils.class.getSimpleName()).build();

    @BindingAdapter("message")
    public static void setMessage (TextView textView, MiPushMessage message) {
        StringBuilder miInternalExtraBuilder = new StringBuilder();
        StringBuilder mptExtraBuilder = new StringBuilder();
        for (String key : message.getExtra().keySet()) {
            StringBuilder builder = key.startsWith(Constants.EXTRA_MIPUSHTESTER_PREFIX) ?
                    mptExtraBuilder : miInternalExtraBuilder;
            builder.append(key);
            builder.append("=");
            builder.append(message.getExtra().get(key));
            builder.append("\n");
        }

        String miuiVersion = "Unkonwn";
        try {        
            Method methodGetString = Build.class.getDeclaredMethod("getString", String.class);
            methodGetString.setAccessible(true);
            miuiVersion = methodGetString.invoke(null, "ro.miui.ui.version.name").toString();
        } catch (Exception e) {
            logger.e("Unable to get property", e);
        }

        textView.setMovementMethod(new ScrollingMovementMethod());
        textView.setText(textView.getContext().getString(R.string.detail,
                message.getMessageId(),
                mptExtraBuilder.toString(),
                miInternalExtraBuilder.toString(),
                Boolean.toString(message.getPassThrough() == 1),
                new GsonBuilder().setPrettyPrinting().create().toJson(message),
                Build.BRAND,
                Build.PRODUCT,
                miuiVersion));
    }
}

File: app/src/main/java/moe/yuuta/mipushtester/push/MessageDetailBindingUtils.kt
package moe.yuuta.mipushtester.push

import android.os.Build
import android.text.method.ScrollingMovementMethod
import android.widget.TextView
import androidx.databinding.BindingAdapter
import com.elvishew.xlog.XLog
import com.google.gson.GsonBuilder
import com.xiaomi.mipush.sdk.MiPushMessage
import moe.yuuta.common.Constants
import moe.yuuta.mipushtester.R

object MessageDetailBindingUtils {
    private val logger = XLog.tag(MessageDetailBindingUtils::class.simpleName).build()

    @JvmStatic
    @BindingAdapter("message")
    fun setMessage (textView: TextView, message: MiPushMessage) {
        val miInternalExtraBuilder = StringBuilder()
        val mptExtraBuilder = StringBuilder()
        for (key in message.extra.keys) {
            val builder = if (key.startsWith(Constants.EXTRA_MIPUSHTESTER_PREFIX))
                    mptExtraBuilder else miInternalExtraBuilder
            builder.append(key)
            builder.append("=")
            builder.append(message.extra.get(key))
            builder.append("\n")
        }

        var miuiVersion = "Unkonwn"
        try {
            val methodGetString = Build::class.java.getDeclaredMethod("getString", String::class.java)
            methodGetString.isAccessible = true
            miuiVersion = methodGetString.invoke(null, "ro.miui.ui.version.name").toString()
        } catch (e: Exception) {
            logger.e("Unable to get property", e)
        }

        textView.movementMethod = ScrollingMovementMethod()
        textView.text = textView.context.getString(R.string.detail,
                message.messageId,
                mptExtraBuilder.toString(),
                miInternalExtraBuilder.toString(),
                if (message.passThrough == 1) "true" else "false",
                GsonBuilder().setPrettyPrinting().create().toJson(message),
                Build.BRAND,
                Build.PRODUCT,
                miuiVersion)
    }
}
